doc/release-notes.md

```diff
@@ -60,6 +60,14 @@ From Bitcoin Core 0.20.0 onwards, macOS versions earlier than 10.12 are no
 longer supported. Additionally, Bitcoin Core does not yet change appearance
 when macOS "dark mode" is activated.
 
+The node's known peers are persisted to disk in a file called `peers.dat`. The
+format of this file has been changed in a backwards-incompatible way in order to
+accommodate the storage of Tor v3 and other BIP155 addresses. This means that if
+the file is modified by 0.21.0 or newer then older versions will not be able to
+read it. Those old versions, in the event of a downgrade, will log an error
+message that deserialization has failed and will continue normal operation
+as if the file was missing, creating a new empty one. (#19954)
```

Comment:
"if the file is modified by 0.21.0 or newer then older versions will not be able to read it. Those old versions, in the event of a downgrade, will log an error message that deserialization has failed and will continue normal operation"

I don't see where this is enforced in 0.20 or earlier versions. If I'm understanding the v0.20 code correctly, then it'll try to unserialize the file, read garbage for the CAddress objects because the format has changed, but then not necessarily fail.

Reply:
You are right!

A dirty hack like the following would trick `<=0.20.1` to always fail deserialization of files in format `>=Format::V3_BIP155` with `Incorrect keysize in addrman deserialization`

```diff
diff --git i/src/addrman.h w/src/addrman.h
index b4089dc89..9f6614cd1 100644
--- i/src/addrman.h
+++ w/src/addrman.h
@@ -326,19 +326,19 @@ public:
     void Serialize(Stream& s_) const
     {
         LOCK(cs);
 
         // Always serialize in the latest version (currently Format::V3_BIP155).
 
         OverrideStream<Stream> s(&s_, s_.GetType(), s_.GetVersion() | ADDRV2_FORMAT);
 
         s << static_cast<uint8_t>(Format::V3_BIP155);
-        s << ((unsigned char)32);
+        s << ((unsigned char)33);
         s << nKey;
         s << nNew;
         s << nTried;
 
         int nUBuckets = ADDRMAN_NEW_BUCKET_COUNT ^ (1 << 30);
         s << nUBuckets;
         std::map<int, int> mapUnkIds;
         int nIds = 0;
         for (const auto& entry : mapInfo) {
@@ -406,18 +406,21 @@ public:
             // Add ADDRV2_FORMAT to the version so that the CNetAddr and CAddress
             // unserialize methods know that an address in addrv2 format is coming.
             stream_version |= ADDRV2_FORMAT;
         }
 
         OverrideStream<Stream> s(&s_, s_.GetType(), stream_version);
 
         unsigned char nKeySize;
         s >> nKeySize;
+        if (format >= Format::V3_BIP155) {
+            --nKeySize;
+        }
         if (nKeySize != 32) throw std::ios_base::failure("Incorrect keysize in addrman deserialization");
         s >> nKey;
         s >> nNew;
         s >> nTried;
         int nUBuckets = 0;
         s >> nUBuckets;
         if (format >= Format::V1_DETERMINISTIC) {
             nUBuckets ^= (1 << 30);
         }
```

Reply:
Yes, that would work. It's essentially repurposing the keysize field in the peers.dat file as a "non forward-compatible version number" (i.e. if a newer version of the software writes a higher number here, then an older version of the software wouldn't be able to read that file).

I believe that's what you wanted to achieve with the code here: https://github.com/bitcoin/bitcoin/pull/19954/files#diff-164bd9e2e30f54d0a79eb7cc372309e2f2155edc6c3f051290ab078f03f6a771R396, but of course that only applies for v0.21 onwards, for file versions higher than `Format::V3_BIP155`. That code means that we now have no way to make a downgrade-safe change to the peers.dat file version. Any bump to the format number means that older software versions can't read the file. Note that we have made downgrade-safe version changes before (I think both v0-v1 and v1-v2 were both downgrade-safe), and it would be nice to keep this as a possibility.

Repurposing the keysize field and reverting the change to the format field seems ok and would almost be turning them into major (non-forward-compatible) and minor (forward-compatible) version numbers. It should go without saying that we always want to have backward-compatibility, so peers.dat files written by old versions are readable by new versions.

Obviously we should get any changes made before the 0.21 release.

Reply:
See if this will make sense to you: https://github.com/bitcoin/bitcoin/pull/20284